### PR TITLE
circulation: load legacy transactions

### PR DESCRIPTION
**Pull request by BadrAly**
* Creates a CLI to load prepared Virtua circulation transactions into
  RERO ILS. Three transaction types are allowed: checkouts, requests and
  fines.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Why are you opening this PR?

https://tree.taiga.io/project/rero21-reroils/us/2011?milestone=289558

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by Garfield-fr — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

File: rero_ils/modules/loans/cli.py [hunk 1]
```diff
@@ -38,12 +38,151 @@
...
+        fields = ['note', 'type', 'patron', 'status', 'organisation',
+                  'total_amount', 'creation_date']
+
+    missing_fields = []
+    for field in fields:
+        try:
+            transaction[field]
+        except Exception as error:
+            missing_fields.append(field)
+    return missing_fields
```
Comment (zannkukai) [inline rero_ils/modules/loans/cli.py:72]:
```python
return [field for field in fields if field not in transaction]
```

**Review by zannkukai — COMMENTED** [review]
> Sound confusing for me to have some legacy load method from Virtua into the RERO project.
In the best of the world, I think this CLI should be extracted from RERO-ILS github project to be integrated into a possible "VirtuaToReroMigration" project

@zannkukai I agree with you, we create a new project `MigrateToReroILS` when we have the first `Alma` library :-)

**Review by benerken — APPROVED** [review]

**Review by iGormilhit — COMMENTED** [review]
Commit message proposition:

```
circulation: load legacy transactions

* Creates a CLI to load prepared Virtua circulation transactions into
  RERO ILS. Three transaction types are allowed: checkouts, requests and
  fines.
  ```

File: rero_ils/modules/loans/cli.py [hunk 2]
```diff
@@ -38,12 +38,145 @@
...
+@click.option('-d', '--debug', 'debug', is_flag=True, default=False)
+@click.argument('infile', type=click.File('r'))
+@with_appcontext
+def load_virtua_transactions(infile, transaction_type, verbose, debug):
+    """Load Virtua circulation transactions.
+
+    infile: Json Virtua transactions file
+    transaction_type: Transaction type either checkout, request or fine.
+    """
+    file_data = json.load(infile)
```
Comment (rerowep) [inline rero_ils/modules/loans/cli.py:105]: If the input files could be big it would be preferable to implement lazy loading.

Comment (rerowep) [reply rero_ils/modules/loans/cli.py:105]: Save of problematic records would be nice.

**Review by zannkukai — APPROVED** [review]

File: rero_ils/modules/loans/cli.py [hunk 3]
```diff
@@ -38,10 +39,166 @@
...
+    """Load Virtua circulation transactions.
+
+    infile: Json Virtua transactions file.
+    transaction_type: Transaction type either checkout, request or fine.
+    :param lazy: lazy reads file.
+    :param save_errors: save error records to file.
+    """
+    # TODO: move this method and other similar methods to a new project that
+    # deals with loading data from other ILS into REROILS.
+    if save_errors:
```
Comment (rerowep) [inline rero_ils/modules/loans/cli.py:113]: It is better to log the error and save the problematic record to a json file for later treatment as it is done for the loading of the other fixtures.

**Review by rerowep — APPROVED** [review]

**Review by iGormilhit — APPROVED** [review]

File: rero_ils/modules/loans/cli.py [hunk 4]
```diff
@@ -38,10 +39,167 @@
...
 from ..patrons.api import Patron, PatronsSearch
 from ..utils import extracted_data_from_ref, get_ref_for_pid, \
-    get_schema_for_resource
+    get_schema_for_resource, read_json_record
+
+
+def check_missing_fields(transaction, transaction_type):
+    """Return list of missing fields for a given transaction type.
+
+    transaction: the json transaction record.
```
Comment (jma) [inline rero_ils/modules/loans/cli.py:52]: Should be: `:param`, see https://thomas-cokelaer.info/tutorials/sphinx/docstring_python.html for more details. Please check similar cases.

File: rero_ils/modules/loans/cli.py [hunk 5]
```diff
@@ -38,10 +39,167 @@
...
+
+def build_loan_record(transaction, transaction_type, item):
+    """Build the loan record before inserting into db.
+
+    transaction: the json transaction record.
+    transaction_type: type of transaction.
+    item: the item record.
+    """
+    if transaction_type == 'checkout':
+        transaction['state'] = 'ITEM_ON_LOAN'
```
Comment (jma) [inline rero_ils/modules/loans/cli.py:78]: The alternative can be to have a dict with new fields and use the `update` dict method.

File: rero_ils/modules/loans/cli.py [hunk 6]
```diff
@@ -38,10 +39,167 @@
...
+        name, ext = os.path.splitext(infile.name)
+        err_file_name = f'{name}_errors{ext}'
+        error_file = open(err_file_name, 'w')
+        error_file.write('[\n')
+
+    if lazy:
+        records = read_json_record(infile)
+    else:
+        file_data = json.load(infile)
+    text = 'Loading Virtua transactions of type {transaction_type}'.format(
```
Comment (jma) [inline rero_ils/modules/loans/cli.py:123]: you can use `f''` here, please check other cases.

**Review by jma — APPROVED** [review]

**Pull request merged**